internal/cloud/realm/client.go

```diff
@@ -91,5 +91,56 @@ func (c *client) do(method, path string, options api.RequestOptions) (*http.Resp
 		req.Header.Set(api.HeaderAuthorization, "Bearer "+auth)
 	}
 	client := &http.Client{}
-	return client.Do(req)
+	res, err := client.Do(req)
+	if err != nil {
+		return res, err
+	}
+	if res.StatusCode != http.StatusUnauthorized || !options.UseAuth || options.PreventRefresh {
+		return res, err
+	}
+
+	serverError := unmarshalServerError(res).(ServerError)
```

Comment:
if we make it to this point, we know the `err` from `client.Do` is nil but our response status code was not successful, at this point we should _try_ to unmarshal the server error

i think we'll need to update the `unmarshServerError` to more gracefully handle any type of response.  you should check how we actually send this error from our server, but I imagine we have some `content-type` header present to indicate a JSON body.  if we see that, unmarshal the error, check for `InvalidSessionCode` and either refresh auth and retry the request or simply return the server error

if we don't see the `content-type` as JSON, then we should construct a generic error (`errors.New(res.Status)` or something like that) and return that instead.  this is a similar flow to how we do it in [baas-admin-sdk](https://github.com/10gen/baas-admin-sdk/blob/master/src/Client.ts#L199-L238)